_blog/2024-02-21-using-jwt-auth-rust.mdx

```diff
@@ -0,0 +1,267 @@
+
+```rust
+#[derive(Debug, Serialize)]
+struct AuthBody {
+    access_token: String,
+    token_type: String,
+}
+
+impl AuthBody {
+    fn new(access_token: String) -> Self {
+        Self {
+            access_token,
+            token_type: "Bearer".to_string(),
+        }
+    }
+}
+```
+
+Now that we’ve created our `AuthBody`, we can create an endpoint that will take a client ID and secret and verify it. Then it will create a claim, encode it and return it as JSON.
+
+```rust
+use axum::Json;
+use chrono::Utc;
+
+#[derive(Debug, Deserialize)]
+struct AuthPayload {
+    client_id: String,
+    client_secret: String,
+}
+
+async fn authorize(Json(payload): Json<AuthPayload>) -> Result<Json<AuthBody>, AuthError> {
+    // Check if the user sent the credentials
+    if payload.client_id.is_empty() || payload.client_secret.is_empty() {
+        return Err(AuthError::MissingCredentials);
+    }
+    // Here, basic verification is used but normally you would use a database
+    if &payload.client_id != "foo" || &payload.client_secret != "bar" {
+        return Err(AuthError::WrongCredentials);
+    }
+
+    // create the timestamp for the expiry time - here the expiry time is 1 day
+    // in production you may not want to have such a long JWT life
+    let exp = (Utc::now().naive_utc() + chrono::naive::Days::new(1)).timestamp() as usize;
+    let claims = Claims {
+        username: payload.client_id,
+        exp,
+    };
+    // Create the authorization token
+    let token = encode(&Header::default(), &claims, &KEYS.encoding).map_err(
+        |_| AuthError::TokenCreation
+    )?;
+
+    // Send the authorized token
+    Ok(Json(AuthBody::new(token)))
+}
+```
+
+Now that we have a route to generate the JWT, we can create a route to try our token out!
+
+```rust
+async fn protected(claims: Claims) -> String {
+    // Send the protected data to the user
+    format!("Welcome to the protected area, {}!", claims.username)
+}
+```
+
+Now that all of the routes have been written we can hook this all back up to our main function:
+
+```rust
+use axum::{Router, routing::{get, post}};
+
+#[shuttle_runtime::main]
+async fn main() -> shuttle_axum::ShuttleAxum {
+    let router = Router::new()
+        .route("/", get(hello_world))
+        .route("/protected", get(protected))
+        .route("/login", post(authorize));
+
+    Ok(router.into())
+}
+```
+
+### Testing
+
+To test that the API works, we can use `cargo shuttle run` to serve the API locally. 
+
+To test our `/login` endpoint, we’ll send a cURL request to it with the data:
+
+```bash
+curl localhost:8000/login -H 'Content-Type: application/json/' \
+ -d '{"client_id":"foo","client_secret":"bar"}'
+```
+
+When we run this, it should output some JSON containing the JWT and type of token - which should be the `Bearer` type. You can prove that this should be a Bearer token because in `FromRequestParts` we extract from the `Authorization: Bearer ..` header.
```

Comment:
```suggestion
When we run this, it should output some JSON containing the JWT and type of token - which should be the `Bearer` type. This should be a Bearer token because in `FromRequestParts` we extract from the `Authorization: Bearer ...` header.
```